source/compilation/Compilation.cpp

```diff
@@ -458,19 +458,23 @@ void Compilation::createDefinition(const Scope& scope, LookupLocation location,
     // Record that the given scope contains this definition. If the scope is a compilation unit, add
     // it to the root scope instead so that lookups from other compilation units will find it.
     auto targetScope = scope.asSymbol().kind == SymbolKind::CompilationUnit ? root.get() : &scope;
+    bool redefined = false;
     std::tuple key(def->name, targetScope);
     if (auto it = definitionMap.find(key); it != definitionMap.end()) {
         reportRedefinition(scope, *def, *it->second);
-        return;
+        redefined = true;
     }
 
-    auto [it, inserted] = definitionMap.emplace(key, std::move(def));
-    ASSERT(inserted);
+    auto [it, inserted] = definitionMap.insert_or_assign(key, std::move(def));
+    if (!redefined) {
```

Comment:
I think this ASSERT is no longer doing anything useful; either the entry is inserted or it's not because of an existing entry -- you're really only checking that the hashmap implements its own interface correctly which can be safely assumed I think.

Reply:
I agree, 

> Maybe the new warning could be called something like `-Wduplicate-definition`

I'm doing that, duplicating `reportRedefinition()` into `reportDuplicateDefinition()`, and while searching for all the calls of the former, it brings up the question of what happens if I have a module redefining a UDP, or vice versa. It could happen in theory (although I specifically don't need this specific case).

I also returned `redefinition` to be an error as before.

Reply:
I think if nobody wants it right now then we should leave it an error. Always better to start out with tight restrictions and relax them if needed, IMO.

Reply:
The question is whether to leave the old `-Wredefinition` as a warning that gets promoted to an error by default (so it can be changed back to warning/ignore by the user), or return it to its original state, where it was a real error that couldn't be overridden.
The new `-Wduplicate-definition` is created as a warning and promoted to an error by default, as was  `-Wredefinition`  in this PR until now.

Reply:
You should return the old `Redefinition` diagnostic to an error.